### PR TITLE
Add Logic for Using TDM and Async Loads and Stores

### PR DESCRIPTION
**Pull request by alex-breslow-amd**
## Motivation

Run fast on gfx1250

## Technical Details

Adds simple copy kernels using TDM and async/loads and stores

## Test Plan

Code has correctness testing, but I need to add more corner case testing

## Test Result

Tests pass on a hardware simulator for both variants:
./TransferBench cmdline 1024K "1 4 G0 T0 G0"
./TransferBench cmdline 1024K "1 4 G0 L0 G0"

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -6,7 +6,15 @@
 ROCM_PATH ?= /opt/rocm
 CUDA_PATH ?= /usr/local/cuda
 MPI_PATH  ?= /usr/local/openmpi
+# pip ROCm wheels ship bin/amdclang++ but often omit bin/amdllvm (which that stub execs).
```
Comment (alex-breslow-amd) [inline Makefile:9]: Logic for running building and running this in a venv on WSL

**Comment by alex-breslow-amd** [comment]
Discussed with @AtlantaPepsi and @gilbertlee-amd.  The plan is to merge as is.  There's some cleanup that needs to be performed however before the candidate branch can be released:
1) Add proper gating for gfx1250
2) Handle case where we are on an NVIDIA platform
3) Add tunability for cache hints to TDM/async load store implementation (Let's sync on this @AtlantaPepsi)

Regarding async loads/stores, the current code is just proof of concept.  The waitcnt placement and unrolling isn't optimized.  Feel free to remove if you think it's best.  TBH, I think this code could be rolled into the existing TransferBench kernels.  I was thinking of working on that (actually what I started on before moving to TDM).  TDM feels sufficiently different where this likely isn't warranted.  I concur with Gilbert on that one.

File: src/header/tdm.h [hunk 2]
```diff
@@ -0,0 +1,355 @@
+/*
```
Comment (alex-breslow-amd) [inline src/header/tdm.h:1]: This file is a standard header but was missing from my installation of ROCm, not sure why.  It can probably be removed.  This is standard in CLR:

File: src/header/TransferBench.hpp [hunk 3]
```diff
@@ -5480,6 +5738,132 @@ static bool IsConfiguredGid(union ibv_gid const& gid)
...
+    do {
+      if (startEvent)
+        ERR_CHECK(hipEventRecord(startEvent, stream));
+#if defined(__NVCC__)
+      GpuAsyncTensorOpsStubKernel<<<dim3(blocks), dim3(threads), 0, stream>>>(src, dst, numFloats);
+#else
+      bool usePipelinedTensorOps = (cfg.tdm.pipelined != 0);
+      auto gpuKernel = tensorFlavor ? (usePipelinedTensorOps ? GpuAsyncPipelinedTensorOpsKernel : GpuAsyncTensorOpsKernel) : GpuAsyncLoadStoreKernel;
+      if (rss.numBytes % sizeof(float) != 0)
+        return {ERR_FATAL, "Async tensor executor (TDM): numBytes (%zu) must be a multiple of %zu", rss.numBytes, sizeof(float)};
```
[the remark this answers is not in the document]
Comment (AtlantaPepsi) [reply src/header/TransferBench.hpp:5772]: this will be elevated into TransfersHaveErrors guard

File: src/header/TransferBench.hpp [hunk 4]
```diff
@@ -4902,6 +4963,203 @@ static bool IsConfiguredGid(union ibv_gid const& gid)
...
+    for (size_t i = 0; i < elementsToProcess; i += elements_per_dwordx4_load_store) {
+      asyncLoadX4(srcPtr + i, shmemPtr + i);
+      __builtin_amdgcn_s_wait_asynccnt(0);
+    }
+
+    // Write back from LDS to global
+    for (size_t i = 0; i < elementsToProcess; i += elements_per_dwordx4_load_store){
+      asyncStoreX4(shmemPtr + i, dstPtr + i);
+      __builtin_amdgcn_s_wait_asynccnt(0);
+    }
```
[the remark this answers is not in the document]
Comment (AtlantaPepsi) [reply src/header/TransferBench.hpp:5156]: async l/s will undergo further changes

File: src/header/TransferBench.hpp [hunk 5]
```diff
@@ -5480,6 +5738,132 @@ static bool IsConfiguredGid(union ibv_gid const& gid)
...
+      hipDeviceProp_t props{};
+      ERR_CHECK(hipGetDeviceProperties(&props, exeIndex));
+      int const warpSize = props.warpSize;
+      int maxShmem = 0;
+      ERR_CHECK(hipDeviceGetAttribute(&maxShmem, hipDeviceAttributeMaxSharedMemoryPerBlock, exeIndex));
+      maxShmem = std::min<int>(maxShmem, cfg.tdm.maxLDSBytes);
+      int kWavesPerWorkgroup = threads / warpSize;
+      int pipelineDepth = usePipelinedTensorOps ? 2 : 1;
+      int numFloatsPerTile = std::max<int>(1, maxShmem / (sizeof(float) *  kWavesPerWorkgroup * pipelineDepth));
+      unsigned int shmemBytes = numFloatsPerTile * kWavesPerWorkgroup * sizeof(float) * pipelineDepth;
```
[the remark this answers is not in the document]
Comment (AtlantaPepsi) [reply src/header/TransferBench.hpp:5784]: to be further validated in next PR
